### PR TITLE
add column to persist workflow attributes

### PR DESCRIPTION
**Pull request by chamilaadhi**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java [hunk 1]
```diff
@@ -59,12 +61,13 @@ public void addWorkflowEntry(Workflow workflow) throws APIMgtDAOException {
...
                 prepStmt.setString(5, workflow.getExternalWorkflowReference());
+                prepStmt.setString(6, WorkflowUtils.mapTojsonString(workflow.getAttributes()));
 
                 prepStmt.execute();
                 connection.commit();
             } catch (SQLException ex) {
                 connection.rollback();
-                throw new APIMgtDAOException(ex);
+                throw new APIMgtDAOException(ex);           
             } finally {
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: Shall we log and throw the exception?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java [hunk 2]
```diff
@@ -192,6 +200,8 @@ public String getExternalWorkflowReferenceForApplication(String appId) throws AP
             ps.setString(2, APIMgtConstants.WorkflowConstants.WF_TYPE_AM_APPLICATION_CREATION);
             try (ResultSet rs = ps.executeQuery()) {
                 workflow = this.createWorkflowFromResultSet(rs);                
+            } catch (ParseException e) {
+                throw new APIMgtDAOException(e);
             }
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: Here also we need to log and throw the exception. Shall we change these?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java [hunk 3]
```diff
@@ -124,14 +127,16 @@ public Workflow retrieveWorkflow(String workflowReference) throws APIMgtDAOExcep
             ps.setString(1, workflowReference);
             try (ResultSet rs = ps.executeQuery()) {
                 workflow = this.createWorkflowFromResultSet(rs);
+            } catch (ParseException e) {
+                throw new APIMgtDAOException(e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: same here log and throw

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java [hunk 4]
```diff
@@ -192,6 +200,8 @@ public String getExternalWorkflowReferenceForApplication(String appId) throws AP
             ps.setString(2, APIMgtConstants.WorkflowConstants.WF_TYPE_AM_APPLICATION_CREATION);
             try (ResultSet rs = ps.executeQuery()) {
                 workflow = this.createWorkflowFromResultSet(rs);                
+            } catch (ParseException e) {
+                throw new APIMgtDAOException(e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: same here log and throw

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/WorkflowUtils.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
+/*
+ *   Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/WorkflowUtils.java]: year needs to be corrected

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/WorkflowUtils.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                    return new LinkedList();
+                }
+
+                @Override
+                public Map createObjectContainer() {
+                    return new HashMap();
+                }
+            };
+            JSONParser parser = new JSONParser();
+            Object obj;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/WorkflowUtils.java]: we can have this in same line

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/WorkflowUtils.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        JSONObject jsonObj = new JSONObject();
+        if (map != null && !map.isEmpty()) {
+            for (Map.Entry<String, String> entry : map.entrySet()) {
+                jsonObj.put(entry.getKey(), entry.getValue());
+            }
+
+            return jsonObj.toJSONString();
+        }
+        return null;
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/WorkflowUtils.java]: additional space?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java (hunk 2 above)
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: done

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java (hunk 4 above)
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: done

**Pull request merged**